File: src/main/java/io/alchemynft/attestation/CompressedMsgSignature.java
package io.alchemynft.attestation;

import java.nio.charset.StandardCharsets;
import org.bouncycastle.crypto.AsymmetricCipherKeyPair;
import org.bouncycastle.crypto.params.AsymmetricKeyParameter;
import org.bouncycastle.util.encoders.Hex;
import org.tokenscript.attestation.core.AttestationCrypto;
import org.tokenscript.attestation.core.SignatureUtility;

public class CompressedMsgSignature implements Signature {
  private static final String TYPE_OF_SIGNATURE = "compressed";
  private final byte[] rawSignature;
  private final String messagePrefix;
  private final String messagePostfix;

  public CompressedMsgSignature(byte[] rawSignature) {
    this(rawSignature, "", "");
  }

  public CompressedMsgSignature(byte[] rawSignature, String messagePrefix, String messagePostfix) {
    this.messagePrefix = messagePrefix;
    this.messagePostfix = messagePostfix;
    this.rawSignature = rawSignature;
  }

  public CompressedMsgSignature(AsymmetricCipherKeyPair keys, byte[] unprocessedMsg) {
    this(keys, unprocessedMsg, "", "");
  }

  /**
   * Constructs a compressed signature of the format @messagePrefix concatenated with Keccak(@unprocessedMsg) concatenated with @messagePostfix.
   */
  public CompressedMsgSignature(AsymmetricCipherKeyPair keys, byte[] unprocessedMsg, String messagePrefix, String messagePostfix) {
    this.messagePrefix = messagePrefix;
    this.messagePostfix = messagePostfix;
    this.rawSignature = sign(keys, unprocessedMsg);
  }

  protected byte[] sign(AsymmetricCipherKeyPair keys, byte[] unprocessedMsg) {
    return SignatureUtility.signWithEthereum(processMessage(unprocessedMsg), keys.getPrivate());
  }

  @Override
  public byte[] getRawSignature() {
    return rawSignature;
  }

  @Override
  public String getTypeOfSignature() {
    return TYPE_OF_SIGNATURE;
  }

  byte[] processMessage(byte[] unprocessedMsg) {
    byte[] hashedUnprocessedMsg = AttestationCrypto.hashWithKeccak(unprocessedMsg);
    String hexEncodedHashedMsg = "0x" + Hex.toHexString(hashedUnprocessedMsg).toUpperCase();
    String stringMsgToSign =  messagePrefix + hexEncodedHashedMsg + messagePostfix;
    return stringMsgToSign.getBytes(StandardCharsets.UTF_8);
  }

  @Override
  public boolean verify(byte[] unprocessedMsg, AsymmetricKeyParameter verificationKey) {
    return SignatureUtility.verifyEthereumSignature(processMessage(unprocessedMsg), rawSignature, verificationKey);
  }
}

File: src/main/javascript/crypto/src/libs/CompressedMsgSignature.ts
import {AbstractSignature} from './AbstractSignature';
import {KeyPair} from "./KeyPair";
import {ethers} from "ethers";
import {SignatureUtility} from "./SignatureUtility";
import {AttestationCrypto} from "./AttestationCrypto";
import {hexStringToArray, stringToArray, uint8tohex} from "./utils";

export class CompressedMsgSignature extends AbstractSignature {
    private static TYPE_OF_SIGNATURE = "compressed";
    private messagePrefix: string;
    private messagePostfix: string;

    public fromRawSignature(rawSignature: string) {
        this.fromRawSignatureAndPrefix(rawSignature, "", "");
    }

    public fromRawSignatureAndPrefix(rawSignature: string, messagePrefix:string, messagePostfix:string) {
        this.messagePrefix = messagePrefix;
        this.messagePostfix = messagePostfix;
        this.rawSignature = rawSignature;
    }

    public fromMessage(keys: KeyPair, unprocessedMsg: Uint8Array) {
        this.fromKeyMessagePrefix(keys, unprocessedMsg, "", "");
    }

    public fromKeyMessagePrefix(keys: KeyPair, unprocessedMsg: Uint8Array, messagePrefix:string, messagePostfix:string) {
        this.messagePrefix = messagePrefix;
        this.messagePostfix = messagePostfix;
        this.rawSignature = this.sign(keys, unprocessedMsg);
    }


    public fromSignature(signature: string) {
        this.signatureAndType(signature, CompressedMsgSignature.TYPE_OF_SIGNATURE);
    }

    protected sign(keys: KeyPair, unprocessedMsg: Uint8Array):string {
        return keys.signRawBytesWithEthereum(this.processMessage(unprocessedMsg));
    }

    public verify(unprocessedMsg: Uint8Array, verificationKey: KeyPair): boolean {
        return verificationKey.verifyBytesWithEthereum(this.processMessage(unprocessedMsg), this.rawSignature);
    }

    public processMessage( unprocessedMsg: Uint8Array): number[] {

        let hashedUnprocessedMsg: Uint8Array = AttestationCrypto.hashWithKeccak(unprocessedMsg);
        let hexEncodedHashedMsg = "0x" + uint8tohex(hashedUnprocessedMsg).toUpperCase();
        let stringMsgToSign:string =  this.messagePrefix + hexEncodedHashedMsg + this.messagePostfix;
        return stringToArray(stringMsgToSign);
    }

    public getTypeOfSignature(): string {
        return CompressedMsgSignature.TYPE_OF_SIGNATURE;
    }

}